To the Editor,

I hope you will consider the attached manuscript, titled "Metagenomic sequencing of soil cores shows taxonomic partitioning of root placement by depth in a prairie plant community," for publication in New Phytologist as a Full Paper.

This work uses a next-generation sequencing strategy, metagenomic barcoding of the ITS2 spacer of ribosomal genes, to identify each plant species present at varying depths in mixed root samples from a prairie restoration. Root identification has traditionally required prohibitive numbers of DNA extractions from individual root segments, tedious and error-prone identification from physical root characteristics, or highly destructive methods such as trenching of sampling sites in order to trace roots back to the plant crown. To move beyond these methods and obtain high coverage of the species present in hundreds of samples at once, our metabarcoding approach uses microfluidic technology (i.e., Fluidigm’s Access Array) to rapidly produce high-quality PCR barcoded tags that can then be sequenced using the latest platform of choice (at the time of our study, Illumina’s MiSeq V3) and processed using tools designed for metagenomic analyses (Pandaseq, VSEARCH), thus opening the door to rapid and highly repeatable characterization of root species present in diverse plant communities. We have used it to identify apparent taxonomic partitioning by depth, with grasses dominant above ground and in the deep soil, but shallow soil layers dominated by forb roots.

* The work addresses the following questions: How do plant species in a managed prairie restoration arrange themselves in the soil profile? Is the observed arrangement consistent with taxonomic partitioning of niche space, especially across depths?

* The work advances our current understanding of plant science: We demonstrate consistent soil exploration strategies within functional groups and predictable consequences of interactions between roots of different taxa. By allowing confident identification of species without physically separating individual root segments or tracing roots to their crowns, we open the possibility of much finer-grained hypothesis tests about root interactions.

* The work is important and timely: It exploits recent advances in environmental sequencing and bioinformatics technologies to test niche partitioning hypotheses at previously infeasible scales. We use the method to evaluate how the spatial arrangement of species affects an ecosystem’s potential to sequester carbon, a topic of current interest for climate and ecological restoration policies.

Best wishes,

Evan
