
## Results and what Discussion there currently is

The sequencing run returned 1286163 plant ITS2 reads, of which 956812 were successfully end-paired, quality filtered, and assigned back from barcodes to samples. These clean reads contained 675898 unique sequences, of which 635969 were observed only once in the dataset (singletons). Of the 39929 sequences observed at least twice, 1537 were identified as probable PCR chimeras by `vsearch`. These chimeric sequences accounted for a total of 5643 reads, or 1.8% of the raw dataset. A further 1854 sequences were identified as incomplete ITS2 regions by `ITSx`, leaving a total of 36142 sequences to be clustered into OTUs. 

We observed more reads from grasses at increasing depths, both relative to other functional types (@Fig:famdepthrel) and in absolute abundance (@Fig:famdepthabs). Forbs from the Asteraceae declined somewhat with depth, while reads from legumes were consistent across depths.

When grouped by genus, additional differences between forbs were visible (@Fig:gendepthrel, @Fig:gendepthabs): Within the Asteraceae, Ratidida (fibrous rooted) and Siplhium (tap-rooted) appear to trade off, with Ratidida less prevalent at middle depths than either shallow or deep layers while Silphium was rare at the surface but most common at middle depths. The fibrous-rooted genera Coreopsis, Solidago, Symphotrichum, and the shallow-taprooted Taraxacum, all declined with depth. All genera of grasses were generally higher in abundance at deeplayers than near the surface, but Panicum, Elymus, and Sorghastrum had their greatest abundance at 75 cm while Andropogon and Schizachrium both continued to increase all the way to 100 cm.

However, some caution is necessary when interpreting these identities. When multiple species from the same genus were detected, they tended to co-occur in the same samples (@Fig:spheatall; note the vertical stripes within each grouped genus). Since most species within each of the genera known from the site are very similar to each other in functional type and growth habit, it is possible that this indicates environmental filtering, where related species tend to grow in the same locations. Or, more parsimoniously, these "multiple species" may be bioinformatic artifacts, either from sequencer error or from genuine biological variation, and all or most of the reads in these groups actually derive from same taxa. Replotting the same heatmap after removing the rarest taxa (sequences assigned to species with mean abundance less than 1% per sample) reduces this effect (@Fig:spheatcommon) but does not remove it entirely. Furthermore, note that the lines for species from genera represented by multiple species in the aboveground checklist (Rudbeckia, Silphium, Solidago) are less similar to each other than those from genera that have only one representative in the checklist. Some refinement of the barcoding pipeline will probably be needed before our taxonomic assignments are trustworthy below the genus level.