
## Results

The MiSeq run and sequence cleanup produced ample, high-quality sequences for the planned anlysis: The raw file contained 1286163 plant ITS2 reads, of which 730235 were successfully end-paired, quality filtered, and assigned back from barcodes to samples. These clean reads contained 494505 unique sequences, of which 459986 were observed only once in the dataset (singletons). Of the 34519 sequences observed at least twice, 2219 were identified as probable or borderline PCR chimeras by `vsearch`. These chimeric sequences accounted for a total of 8526 reads, or ~1.2% of the raw dataset. A further 75 sequences were identified as incomplete or undetectable ITS2 regions by `ITSx`, leaving a total of 32225 ITS2 sequences to be clustered at 99% similarity into 1347 OTUs, which we then used as reference sequences at 99% similarity to map the 730235 reads (singletons included) of the cleaned sequence file, for a final sequence-by-sample table of 576650 reads. Sample coverage was excellent, with 110 of the 120 mixed root samples having enough reads (1355-8998) for confident analysis. There was no apparent change in total read count between samples from differing depths.

After obtaining best-match barcode identities for each of the 1347 OTUs clustered at 99%, we collapsed OTUs to phylotypes by combining groups of OTUs that all yielded the same taxon as their top BLAST hit. When we collapsed at the species level, we obtained 158 phylotypes that were identified as coming from 70 genera across 16 families, which is notably higher than the 32 species, 22 genera, and 6 families known from site vegetation surveys [@Tbl:splist]. Many samples contained high read counts from several different `species` of genera that have only one known species at the site [Which figure shows this best? Maybe @Fig:mock_control; @Fig:spike_control]. From this we infer that, at least with current databases, ITS2 barcode identities are may be more reliable at the genus level than the species level.

We observed more reads from grasses at increasing depths, both relative to other functional types (@Fig:famdepthrel) and in absolute abundance (@Fig:famdepthabs). Forbs from the Asteraceae declined somewhat with depth, while reads from legumes were consistent across depths.

When grouped by genus, additional differences between forbs were visible (@Fig:gendepthrel, @Fig:gendepthabs): Within the Asteraceae, Ratidida (fibrous rooted) and Siplhium (tap-rooted) appear to trade off, with Ratidida less prevalent at middle depths than either shallow or deep layers while Silphium was rare at the surface but most common at middle depths. The fibrous-rooted genera Coreopsis, Solidago, Symphotrichum, and the shallow-taprooted Taraxacum, all declined with depth. All genera of grasses were generally higher in abundance at deeplayers than near the surface, but Panicum, Elymus, and Sorghastrum had their greatest abundance at 75 cm while Andropogon and Schizachrium both continued to increase all the way to 100 cm.

...talk about weighting by total root mass here; likely that all species have greatest mass at the surface but that forbs win for overall total while grasses have small but ecologically important deep pools.

<!-- PROBABLY DELETE ME
However, some caution is necessary when interpreting these identities. When multiple species from the same genus were detected, they tended to co-occur in the same samples (@Fig:spheatall; note the vertical stripes within each grouped genus). Since most species within each of the genera known from the site are very similar to each other in functional type and growth habit, it is possible that this indicates environmental filtering, where related species tend to grow in the same locations. Or, more parsimoniously, these "multiple species" may be bioinformatic artifacts, either from sequencer error or from genuine biological variation, and all or most of the reads in these groups actually derive from same taxa. Replotting the same heatmap after removing the rarest taxa (sequences assigned to species with mean abundance less than 1% per sample) reduces this effect (@Fig:spheatcommon) but does not remove it entirely. Furthermore, note that the lines for species from genera represented by multiple species in the aboveground checklist (Rudbeckia, Silphium, Solidago) are less similar to each other than those from genera that have only one representative in the checklist. Some refinement of the barcoding pipeline will probably be needed before our taxonomic assignments are trustworthy below the genus level.
-->
